src/System.Data.SqlClient/src/System/Data/SqlClient/SqlDataReader.cs

```diff
@@ -2583,85 +2583,136 @@ private T GetFieldValueInternal<T>(int i)
 
         private T GetFieldValueFromSqlBufferInternal<T>(SqlBuffer data, _SqlMetaData metaData)
         {
-            Type typeofT = typeof(T);
-            if (_typeofINullable.IsAssignableFrom(typeofT))
+            // this block of type specific shortcuts uses RyuJIT jit behaviours to achieve fast implementations of the primative types
+            // RyuJIT will be able to determine at compilation time that the typeof(T)==typeof(<primative>) options are constant
+            // and be able to remove all implementations which cannot be reached. this will eliminate non-specialized code for value types
```

Comment:
@Wraith2 
Have you checked the assumption?
`dataType` is known at runtime only, so JIT won't be able to do the optimization.

Reply:
The `typeof(T) == typeof(byte)` should eliminate the branches; it will then leave behind `dataType == typeof(byte)`

I assume that's why the double checks with `typeof(T)` being first?

Reply:
It does, i answered in the main thread not this branch but sharplab confirms the elimination because one of the conditions is constantly false.

Reply:
@Wraith2 
Hmm.. yes, you are right. 
Wouldn't then be more efficient to compare directly with `StorageType` enum rather than creating an instance of `Type` just for comparison? Types are cached but the cash is finite and you could save a call.

Reply:
It does the storage type checks first and if a direct match is found will return the value as directly as possible, if there's a mismatch then it falls back to the original path of doing the type check. So yes it's better to do what you said and it does.

I believe since it's generic that typeof(T) is low cost because the jit can fold in the type handle directly not requiring a method call. The other part is a readonly static to avoid the lookup cost.